# Лабораторная работа 1. Методы агрегации и группировки данных

**Вариант 5.** Набор данных `visits.csv` — Данные пользовательских сессий магазина

**Цель работы:** овладение навыками агрегации и группировки табличных
данных для преобразования детальных записей в структурированные
аналитические отчёты.

In [ ]:
import pandas as pd

url = 'https://raw.githubusercontent.com/versanivc/LR/refs/heads/main/visits.csv'

# Чтение таблицы с разделителем ;
df = pd.read_csv(url, sep=';')

## 2. Вывод первых 20 строк

С помощью метода `head()` выводим первые 20 строк набора данных, чтобы составить общее впечатление о данных.

In [ ]:
# Вывод первых 20 строк
df.head(20)

,User_Id,Region,Device,Channel,Session_Start,SESSION_End
0,"9,81449E+11",United States,iPhone,organic,01.05.2019 2:36,01.05.2019 2:45
1,"2,78966E+11",United States,iPhone,organic,01.05.2019 4:46,01.05.2019 4:47
2,"5,90706E+11",United States,Mac,organic,01.05.2019 14:09,01.05.2019 15:32
3,"3,26434E+11",United States,Android,TipTop,01.05.2019 0:29,01.05.2019 0:54
4,"3,49774E+11",United States,Mac,organic,01.05.2019 3:33,01.05.2019 3:57
5,43958116050,United States,Android,organic,01.05.2019 9:03,01.05.2019 10:08
6,"1,85365E+11",United States,iPhone,organic,01.05.2019 9:37,01.05.2019 10:00
7,"1,00971E+11",United States,Mac,TipTop,01.05.2019 4:39,01.05.2019 4:57
8,"3,70456E+11",United States,iPhone,organic,01.05.2019 14:44,01.05.2019 15:41
9,"1,41838E+11",United States,Mac,FaceBoom,01.05.2019 6:20,01.05.2019 6:54


## 3. Обзор данных - описание столбцов и предметной области

Набор данных содержит информацию о пользовательских сессии магазина. Описание столбцов:

- **User_Id** - уникальный идентификатор пользователя
- **Region** -  страна пользователя
- **Device** -  устройство пользователя
- **Channel** - идентификатор рекламного источника, из которого пришел
пользователь
- **Session_Start** - дата и время начала сессии
- **SESSION_End** - дата и время окончания сессии


Предметная область — электронная коммерция. Данные могут использоваться для анализа поведения пользователей, продолжительности сессий, эффективности рекламных каналов, а также различий в активности пользователей в зависимости от региона и типа устройства.

## 4. Оценка данных с помощью info()

Метод `info()` предоставляет общую информацию о таблице: названия столбцов, типы данных, хранящиеся в них, и количество ненулевых значений в каждом столбце.

In [ ]:
# Общая информация о таблице
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 957 entries, 0 to 956
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   User_Id        957 non-null    object
 1   Region         956 non-null    object
 2   Device         955 non-null    object
 3   Channel        957 non-null    object
 4   Session_Start  955 non-null    object
 5   SESSION_End    955 non-null    object
dtypes: object(6)
memory usage: 45.0+ KB


**Результаты:**
- Всего строк: 957
- Столбцов: 6
- Обнаружены пропуски в столбцах: `Region` (1), `Device` (2), `Session_Start` (2), `SESSION_End` (2),
- Без пропусков: `User_Id`, `Channel`
- Типы данных: `object` (6)

**Проблемы:**
- `Session_Start` определён как `object` вместо даты и времени
- `SESSION_End` определён как `object` вместо даты и времени
- `SESSION_End` записан в верхнем регистре, в отличие от `Session_End`
- Наличие пропусков требует обработки


## 5. Применение describe()

С помощью `describe()` оценим числовые столбцы: посмотрим на средние значения, минимум, максимум, стандартное отклонение и квартили.

In [ ]:
# Статистика по числовым столбцам
df.describe()

,User_Id,Region,Device,Channel,Session_Start,SESSION_End
count,957,956,955,957,955,955
unique,865,2,6,4,820,836
top,"1,79085E+11",United States,iPhone,organic,02.05.2019 9:40,02.05.2019 22:49
freq,4,955,421,612,4,3


С помощью `describe()` оценены числовые столбцы:

- **User_Id:** всего 957 записей, из них 865 уникальных пользователей. Некоторые пользователи посещали сайт несколько раз.
- **Region:** пользователи представлены из 2 стран, при этом почти все записи относятся к United States.
- **Device:** в данных присутствует 6 типов устройств. Наиболее популярное устройство — iPhone, встречается 421 раз.
- **Channel:** используется 4 рекламных канала. Чаще всего пользователи приходят через канал organic — 612 сессий.
- **Session_Start:** зафиксировано 820 уникальных моментов начала сессий, некоторые даты и время повторяются.
- **SESSION_End:** присутствует 836 уникальных значений времени окончания сессий.

## 6. Проверка названий столбцов

Выведем названия столбцов с помощью `df.columns` и проверим, есть ли проблемы: пробелы в названиях, разные языки, непонятные названия.

In [ ]:
# Вывод названий столбцов
print(df.columns.tolist())

['User_Id', 'Region', 'Device', 'Channel', 'Session_Start', 'SESSION_End']


Как видно выше, названия столбцов в целом оформлены корректно: отсутствуют пробелы, большинство столбцов придерживаются единого стиля именования. Однако наблюдается небольшое несоответствие — столбец SESSION_End записан в верхнем регистре, в отличие от остальных названий. После приведения к единому стилю он был приведён к виду Session_End.

In [ ]:
# Переименование отдельных столбцов
df = df.rename(columns={'SESSION_End': 'Session_End'})

# Вывод названий столбцов
print(df.columns.tolist())


['User_Id', 'Region', 'Device', 'Channel', 'Session_Start', 'Session_End']


С помощью метода `rename()` столбец `SESSION_End` был переименован в `Session_End`. Теперь названия столбцов приведены к единому стилю: отсутствуют пробелы, имена оформлены в согласованном формате, а также являются понятными и отражают суть данных.

## 7. Проверка данных на наличие пропусков

Для нахождения пропусков применяется метод `isna()`. Потом результат работы `isna()` передаём методу `sum()`.

In [ ]:
# Подсчёт количества пропусков в каждом столбце
print(df.isna().sum())

User_Id          0
Region           1
Device           2
Channel          0
Session_Start    2
Session_End      2
dtype: int64


В результате проверки обнаружены пропуски в 4 столбцах:
- `Region` - 1 пропуск
- `Device` - 2 пропуска
- `Session_Start` - 2 пропуска
- `Session_End ` - 2 пропуска

Способ обработки пропусков зависит от их типа и значимости столбца:
- Строки без данных в столбцах  (`Session_Start` и `Session_End`) удалены, так как без этой информации запись теряет смысл.
- Пропуски в столбцах (`Region` и `Device`) заполним пропущенные значения наиболее часто встречающимися значениями в таблице. - это позволяет сохранить строки и не исказить общее распределение значений.


In [ ]:
# 1. Удаляем строки, где пропущено начало и конец сессии
df = df.dropna(subset=['Session_Start'])
df = df.dropna(subset=['Session_End'])
# 2. Заполняем столбцы  Регион и Столбец наиболее часто
# встречающимися значениями в таблице
df['Region'] = df['Region'].fillna(df['Region'].mode()[0])
df['Device'] = df['Device'].fillna(df['Device'].mode()[0])


# 4. Проверяем, что пропусков не осталось
print(df.isna().sum())

User_Id          0
Region           0
Device           0
Channel          0
Session_Start    0
Session_End      0
dtype: int64


Пропуски успешно обработаны. Во всех столбцах количество пропусков - 0. Данные готовы к дальнейшей проверке.

## 8. Проверка данных на наличие дубликатов

Проверим явные и неявные дубликаты. Для поиска явных дубликатов используется метод `duplicated()`.

In [ ]:
# Поиск и подсчёт явных дубликатов
print('Количество дубликатов:', df.duplicated().sum())

# Вывод строк-дубликатов
print('\nСтроки-дубликаты:')
display(df[df.duplicated()])

# Удаление дубликатов и обновление индексации
df = df.drop_duplicates().reset_index(drop=True)

# Проверяем, что дубликатов не осталось
print('\nКоличество дубликатов после удаления:', df.duplicated().sum())

Количество дубликатов: 2

Строки-дубликаты:


,User_Id,Region,Device,Channel,Session_Start,Session_End
480,"1,79085E+11",United States,Mac,organic,01.05.2019 7:27,01.05.2019 7:49
481,"1,79085E+11",United States,Mac,organic,01.05.2019 7:27,01.05.2019 7:49



Количество дубликатов после удаления: 0


Для поиска неявных дубликатов — смотрим уникальные значения в текстовых столбцах

In [ ]:
# Поиск неявных дубликатов: смотрим уникальные значения в  столбцах
print('Уникальные значения в столбце "Region":')
print(df['Region'].unique())

print('\nУникальные значения в столбце "Device":')
print(df['Device'].unique())

print('\nУникальные значения в столбце "Channel":')
print(df['Channel'].unique())

Уникальные значения в столбце "Region":
['United States' 'USA']

Уникальные значения в столбце "Device":
['iPhone' 'Mac' 'Android' 'PC' 'MAC' 'IPHONE']

Уникальные значения в столбце "Channel":
['organic' 'TipTop' 'FaceBoom' 'MediaTornado']


Приведено приведение значений столбцов `Region` и `Device` к единому формату с использованием функции `replace()`

In [ ]:

# Приводим значения столбца Region к единому формату:
# заменяем сокращение 'USA' на полное название 'United States'
df['Region'] = df['Region'].replace('USA', 'United States')
# Приводим значения столбца Device к единому стилю написания:
# заменяем некорректные/разные варианты названий устройств на стандартные
df['Device'] = df['Device'].replace({
'MAC': 'Mac','IPHONE': 'iPhone'})
# Проверяем
print(df['Region'].unique())
print(df['Device'].unique())

['United States']
['iPhone' 'Mac' 'Android' 'PC']


**Явные дубликаты.** С помощью метода `duplicated()` было обнаружено 2 повторяющиеся строки (один пользователь с одной датой начала сессии — 01.05.2019 7:27 — встречается дважды). Дубликаты были удалены с помощью метода `drop_duplicates()` с обновлением индексации.

**Неявные дубликаты.** С помощью метода `unique()` в столбцах `Region` и `Device` были выявлены несогласованные записи. В столбце `Region` сокращение `USA` было заменено на полное название `United States` с помощью метода `replace()`. В столбце `Device` значения `MAC` и `IPHONE` были приведены к единому формату (`Mac` и `iPhone` соответственно) также с использованием `replace()`.

После преобразований проверка методом `unique()` показала, что значения стандартизированы. В столбце `Channel` неявных дубликатов обнаружено не было.

## 9. Проверка типов данных

Проверим типы данных в столбцах и при необходимости изменим их, чтобы они соответствовали действительности.

In [ ]:
# Проверка текущих типов данных
print(df.dtypes)

User_Id          object
Region           object
Device           object
Channel          object
Session_Start    object
Session_End      object
dtype: object


Столбцы `Session_Start` и `Session_End` имеют тип `object`, хотя должны быть в формат даты и времени (datetime)

In [ ]:
# Приведение к формату даты и времени
df['Session_Start'] = pd.to_datetime(df['Session_Start'], format='%d.%m.%Y %H:%M')
df['Session_End'] = pd.to_datetime(df['Session_End'], format='%d.%m.%Y %H:%M')
# Проверяем результат
print(df.dtypes)

User_Id                  object
Region                   object
Device                   object
Channel                  object
Session_Start    datetime64[ns]
Session_End      datetime64[ns]
dtype: object


## 10. Выводы по работе

По итогам выполнения контрольной работы был проведён полный цикл предварительного анализа и очистки набора данных `visits.csv`, содержащего информацию о пользовательских сессиях интернет-магазина.

В процессе работы были последовательно решены следующие задачи: загрузка данных с корректным указанием разделителя, первичный визуальный осмотр первых 20 строк, анализ структуры таблицы с помощью `info()` и `describe()`. В ходе анализа выявлены и устранены типичные проблемы исходных данных: некорректные типы данных в столбцах с датой и временем, пропуски значений в столбцах `Region`, `Device`, `Session_Start`, `Session_End`, явные дубликаты строк, а также неявные дубликаты, связанные с несогласованным написанием значений в категориальных переменных (`USA / United States`, `MAC / Mac`, `IPHONE / iPhone`). Дополнительно приведено к единому стилю название столбца `SESSION_End`.

Для обработки пропусков использовались удаление строк с критическими пропусками и заполнение модальными значениями. Дубликаты удалены, а категориальные переменные стандартизированы. Типы данных столбцов `Session_Start` и `Session_End` успешно преобразованы в формат `datetime`.

В результате получен очищенный, структурированный и готовый к дальнейшему анализу набор данных, что подтверждает достижение цели работы и освоение ключевых методов предобработки данных с использованием библиотеки pandas.

##**Группировка данных и построение сводных таблиц**

##Задание 1 : Группировка - `device` и количество рекламных источников каждого типа (`channel`).

In [ ]:
# Группировка по устройству и каналу с подсчетом количества
task1=df.groupby('Device')['Channel'].count()
print(task1)

Device
Android    185
Mac        242
PC         103
iPhone     423
Name: Channel, dtype: int64


По результатам группировки видно, что наибольшее количество рекламных источников (channel) связано с устройствами `iPhone — 423` записи. Это может говорить о том, что пользователи `iPhone` наиболее активно взаимодействуют с рекламными каналами или составляют основную часть аудитории.

На втором месте находятся устройства `Mac — 242` записи, затем `Android — 185` записей. Наименьшее количество наблюдается у пользователей `PC — 103` записи.

Таким образом, можно сделать вывод, что мобильные устройства, особенно продукция `Apple (iPhone и Mac)`, доминируют в выборке и чаще используются для взаимодействия с рекламными источниками. Пользователи ПК проявляют наименьшую активность.

##Задание 2 : Группировка - `region` и количество устройств (`device`).Создать датафрейм. Переименовать столбец с количеством в “сount”. Отсортировать по убыванию столбца “count”.


In [ ]:
task2= (
    df.groupby(['Region','Device'])['Device']
    .count()
    .reset_index(name='count')
    .sort_values(by='count', ascending=False)
)
print(task2)

          Region   Device  count
3  United States   iPhone    423
1  United States      Mac    242
0  United States  Android    185
2  United States       PC    103


Из полученного датафрейма видно, что все устройства относятся к региону `US`. Наибольшее количество пользователей использует устройство `iPhone — 423` записи, что делает его самым популярным устройством в данном регионе.

На втором месте находится `Mac — 242` устройства, далее `Android — 185` устройств. Меньше всего зафиксировано пользователей с устройствами `PC — 103` записи.

Сортировка по убыванию столбца count позволяет быстро определить наиболее популярные типы устройств. Можно сделать вывод, что в регионе `US` преобладают пользователи техники `Apple`, особенно `iPhone`

##Задание 3 :  Сводная таблица (`pivot_table`) - уникальное количество пользователей для каждого устройства (`device`). Отсортировать по убыванию количества


In [ ]:
task3= df.pivot_table(index='Device',
                      values='User_Id',
                      aggfunc='nunique'
                      ).reset_index().sort_values(by='User_Id', ascending=False)
print(task3)

    Device  User_Id
3   iPhone      385
1      Mac      223
0  Android      169
2       PC       90


Наибольшее количество уникальных пользователей используют `iPhone — 385` человек. Это говорит о том, что данное устройство является самым популярным среди пользователей в исследуемой выборке.

На втором месте находится `Mac — 223` уникальных пользователя, затем `Android — 169` пользователей. Наименьшее количество пользователей зафиксировано у устройств `PC — 90` человек.

После сортировки по убыванию видно, что устройства `Apple (iPhone и Mac)` значительно преобладают над остальными устройствами.

##Задание 4 :   Сводная таблица (pivot_table) - уникальное количество пользователей для каждого устройства (device) - столбцы и региона- строки. Отсортировать по возрастанию столбца region.


In [ ]:
pivot_users= df.pivot_table(
    index='Device',
    columns='Region',
    values='User_Id',
    aggfunc='nunique').sort_index()
print(pivot_users)

Region   United States
Device                
Android            169
Mac                223
PC                  90
iPhone             385


Видно, что основная пользовательская активность сосредоточена на устройствах `Apple`, особенно на `iPhone`, что подтверждает преобладание мобильного поведения в электронной коммерции.

##Задание 5 :   Добавить столбец - “Длительность сессии” (расчетный). Создать столбец “Категория длительности” (с помощью категоризации). Выделить минимум 3 категории (низкая, высокая, средняя), фильтрацию для уровня длительности выбрать самостоятельно, аргументировать выбор. Создать группировку: средняя, минимальная, максимальная, медианная длительность сессии по категории длительности и устройству


In [ ]:
df['Session_duration'] = df['Session_End'] - df['Session_Start']
print(df)




         User_Id         Region   Device   Channel       Session_Start  \
0    9,81449E+11  United States   iPhone   organic 2019-05-01 02:36:00   
1    2,78966E+11  United States   iPhone   organic 2019-05-01 04:46:00   
2    5,90706E+11  United States      Mac   organic 2019-05-01 14:09:00   
3    3,26434E+11  United States  Android    TipTop 2019-05-01 00:29:00   
4    3,49774E+11  United States      Mac   organic 2019-05-01 03:33:00   
..           ...            ...      ...       ...                 ...   
948  5,55066E+11  United States   iPhone    TipTop 2019-05-04 23:32:00   
949  7,71716E+11  United States  Android    TipTop 2019-05-04 19:32:00   
950  74352279110  United States   iPhone    TipTop 2019-05-04 00:13:00   
951  6,08683E+11  United States  Android  FaceBoom 2019-05-04 03:33:00   
952  8,75326E+11  United States      Mac   organic 2019-05-04 03:01:00   

            Session_End Session_duration  
0   2019-05-01 02:45:00  0 days 00:09:00  
1   2019-05-01 04:47:00  

Добавлен новый столбец «Длительность сессии» и заполнены данные для категории длительности.
Категории длительности выбраны следующим образом:

*  до 10 минут — низкая длительность;
*   от 10 до 20 минут — средняя;
*  более 20 минут — высокая.

Такое разделение позволяет выделить короткие, стандартные и длительные пользовательские сессии.

In [ ]:
bins=[pd.Timedelta('0 days 00:00:00'), pd.Timedelta('0 days 00:10:00'),
      pd.Timedelta('0 days 00:20:00'), pd.Timedelta('0 days 02:00:00') ] # границы
labels = ['Низкая', 'Средняя', 'Высокая']

df['Duration_Category'] = pd.cut(df['Session_duration'], bins=bins, labels=labels)

task4=df.groupby(['Device','Duration_Category']).agg({'Session_duration':['mean','min','max','median']}).reset_index()
print(task4)

     Device Duration_Category          Session_duration                  \
                                                   mean             min   
0   Android            Низкая 0 days 00:05:19.534883720 0 days 00:01:00   
1   Android           Средняя           0 days 00:14:27 0 days 00:11:00   
2   Android           Высокая 0 days 00:48:35.049504950 0 days 00:21:00   
3       Mac            Низкая           0 days 00:06:00 0 days 00:01:00   
4       Mac           Средняя 0 days 00:15:48.196721311 0 days 00:11:00   
5       Mac           Высокая 0 days 00:46:37.894736842 0 days 00:21:00   
6        PC            Низкая 0 days 00:05:17.142857142 0 days 00:01:00   
7        PC           Средняя 0 days 00:15:46.363636363 0 days 00:11:00   
8        PC           Высокая           0 days 00:46:45 0 days 00:21:00   
9    iPhone            Низкая 0 days 00:04:39.173553719 0 days 00:01:00   
10   iPhone           Средняя 0 days 00:15:06.428571428 0 days 00:11:00   
11   iPhone           Выс

/tmp/ipykernel_14988/1480256983.py:7: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  task4=df.groupby(['Device','Duration_Category']).agg({'Session_duration':['mean','min','max','median']}).reset_index()


В ходе анализа было выявлено, что пользователи PC проводят в сессиях больше времени, особенно в категории высокой длительности. Пользователи iPhone имеют самые короткие сессии в категории низкой длительности. Средние и медианные значения близки друг к другу, поэтому данные распределены достаточно равномерно. Категории длительности были выбраны корректно и позволяют разделить сессии на короткие, средние и длинные.

##Задание 6 :   Добавить столбец - “Длительность сессии” (расчетный). Создать столбец “Категория длительности” (с помощью категоризации). Выделить минимум 3 категории (низкая,высокая, средняя), фильтрацию для уровня длительности выбрать самостоятельно, аргументировать выбор. Отфильтровать набор данных. Выбрать только высокую и низкую длительность сессии + 2 самых не популярных канала (по количеству записей). Создать сводную таблицу: средняя, медианная длительность сессии по каналу и устройству.



In [ ]:
task6=df[df['Duration_Category'].isin(['Низкая','Высокая'])]
print(task6)

         User_Id         Region   Device  Channel       Session_Start  \
0    9,81449E+11  United States   iPhone  organic 2019-05-01 02:36:00   
1    2,78966E+11  United States   iPhone  organic 2019-05-01 04:46:00   
2    5,90706E+11  United States      Mac  organic 2019-05-01 14:09:00   
3    3,26434E+11  United States  Android   TipTop 2019-05-01 00:29:00   
4    3,49774E+11  United States      Mac  organic 2019-05-01 03:33:00   
..           ...            ...      ...      ...                 ...   
943  64474110919  United States   iPhone  organic 2019-05-04 00:24:00   
945  8,61181E+11  United States      Mac  organic 2019-05-04 07:52:00   
946  4,10845E+11  United States  Android   TipTop 2019-05-04 05:26:00   
947  6,86573E+11  United States  Android   TipTop 2019-05-04 17:42:00   
948  5,55066E+11  United States   iPhone   TipTop 2019-05-04 23:32:00   

            Session_End Session_duration Duration_Category  
0   2019-05-01 02:45:00  0 days 00:09:00            Низкая  
1

Ранее в задании 5 был создан столбец «Длительность сессии» (расчетный). Также был добавлен столбец «Категория длительности» с использованием категоризации, где выделены минимум три категории: низкая, средняя и высокая.

Также была проведена фильтрация данных. А именно, были отобраны только записи с высокой и низкой длительностью сессии.

In [ ]:
# Считаем количество записей по каждому каналу
channel_count=df['Channel'].value_counts()
print("Популярность каналов:")
print(channel_count)
# Берем 2 самых НЕпопулярных (с наименьшим количеством)
least_popular=channel_count.tail(2).index.tolist()

print("\n2 самых НЕпопулярных канала:")
print(least_popular)
task6 = df[df['Channel'].isin(least_popular)]
print("\n")
result = pd.pivot_table(
    task6,
    index='Channel',
    columns='Device',
    values='Session_duration',
    aggfunc=['mean', 'median']
)

print(result)

Популярность каналов:
Channel
organic         609
FaceBoom        149
TipTop          142
MediaTornado     53
Name: count, dtype: int64

2 самых НЕпопулярных канала:
['TipTop', 'MediaTornado']


                                  mean                                  \
Device                         Android             Mac              PC   
Channel                                                                  
MediaTornado 0 days 00:20:27.272727272 0 days 00:25:54 0 days 00:09:00   
TipTop       0 days 00:29:57.692307692 0 days 00:26:05 0 days 00:32:30   

                                                median                  \
Device                          iPhone         Android             Mac   
Channel                                                                  
MediaTornado           0 days 00:23:10 0 days 00:13:00 0 days 00:24:00   
TipTop       0 days 00:26:50.769230769 0 days 00:24:30 0 days 00:16:30   

                                              
Device          

Из полученной сводной таблицы видно, что у двух наименее популярных каналов — `TipTop` и `MediaTornado` — заметно различается поведение пользователей в зависимости от устройства. `MediaTornado` в среднем показывает более короткие сессии на `Android` и особенно на `ПК`, тогда как на `Mac` и `iPhone` время выше и более стабильное. При этом медианные значения подтверждают, что на `ПК` и `Android` сессии чаще всего короче, чем на других устройствах.

У `TipTop` общая длительность сессий выше и более равномерная между устройствами, особенно на `ПК`, где среднее время заметно выше остальных платформ. В целом, можно сказать, что `TipTop` удерживает пользователей стабильнее, тогда как `MediaTornado` демонстрирует более слабую вовлечённость, особенно на десктопных устройствах.

##Задание 7 :   Проанализировать среднюю и медианную длительность сессий пользователей в зависимости от рекламного канала, но только для мобильных устройств на Android и PC. Необходимо отфильтровать данные по устройствам, затем построить сводную таблицу с группировкой по каналу и устройству



In [ ]:
# фильтруем только мобильные устройства
task7 = df[df['Device'].isin(['Android','PC'])]
# строим сводную таблицу
result_task7 = pd.pivot_table( task7,
    index='Channel',
    columns='Device',
    values='Session_duration',
    aggfunc=['mean','median']
)
print(result_task7)

                                  mean                            \
Device                         Android                        PC   
Channel                                                            
FaceBoom     0 days 00:28:25.454545454           0 days 00:32:20   
MediaTornado 0 days 00:20:27.272727272           0 days 00:09:00   
TipTop       0 days 00:29:57.692307692           0 days 00:32:30   
organic                0 days 00:32:48 0 days 00:30:13.333333333   

                      median                  
Device               Android              PC  
Channel                                       
FaceBoom     0 days 00:23:00 0 days 00:20:00  
MediaTornado 0 days 00:13:00 0 days 00:09:00  
TipTop       0 days 00:24:30 0 days 00:27:30  
organic      0 days 00:22:00 0 days 00:19:30  


Из полученной сводной таблицы видно, что длительность сессий заметно различается в зависимости от рекламного канала и устройства. Наиболее стабильное поведение пользователей наблюдается у канала `MediaTornado` на `PC`, где средние и медианные значения совпадают и остаются низкими(9 минут), что говорит о коротких и однотипных сессиях.

В каналах `organic`, `FaceBoom` и `TipTop` средняя длительность сессий выше медианы, что указывает на наличие отдельных длинных сессий, увеличивающих среднее значение и свидетельствующих о более разнообразном поведении пользователей. В целом, пользователи `PC` и `Android` демонстрируют схожие паттерны, однако в отдельных каналах (особенно `organi`) наблюдается более высокая вовлечённость и разброс по времени сессий.

##Вывод:
По итогам работы выполнена полная очистка данных: устранены пропуски, дубликаты и несогласованности в названиях, а также исправлены типы данных для столбцов с датой и временем.

С помощью группировок и сводных таблиц выявлено, что большинство сессий приходится на iPhone и органический трафик, а основная активность пользователей сосредоточена в США. Анализ длительности сессий показал, что владельцы iPhone чаще имеют короткие сессии, тогда как пользователи PC проводят время дольше. Наименее популярные каналы (TipTop и MediaTornado) демонстрируют разную вовлечённость в зависимости от устройства, а на мобильных устройствах органический трафик и FaceBoom обеспечивают более высокую длительность взаимодействия.
